### PR TITLE
Add HKDF_HMAC256_L32 and method to negate a private key

### PR DESCRIPTION
**Pull request by jonasschnelli**
This adds a limited implementation of `HKDF` (defined by rfc5869) that supports only HMAC-SHA256  and length output of 32 bytes (will be required for v2 transport protocol).

This PR also includes a method to negate a private key which is useful to enforce public keys starting with 0x02 (or 0x03) (a requirement for the v2 transport protocol). The new `CKey::Negate()` method is pretty much a wrapper around `secp256k1_ec_privkey_negate()`.

Including tests.

This is a subset of #14032 and a pre-requirement for the v2 transport protocol.

### CODE REVIEW

File: src/crypto/hkdf_sha256_32.cpp [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+void CHKDF_HMAC_SHA256_L32::Expand32(const std::string& info, unsigned char* out)
+{
+    // expand a 32byte key (single round)
+    assert(info.size() <= 128);
+    unsigned char data[129];
+    memcpy(&data[0], &info[0], info.size());
+    data[info.size()] = 0x01;
+    CHMAC_SHA256(m_prk, 32).Write(&data[0], info.size() + 1).Finalize(out);
```
Comment (sipa) [inline src/crypto/hkdf_sha256_32.cpp]: You can avoid the copy to `data` if you use `static const char one[1] = {1}; CHMAC_SHA256(m_prk, 32).Write(info.data(), info.size()).Write(one, 1).Finalize(out);` instead.

Comment (jonasschnelli) [reply src/crypto/hkdf_sha256_32.cpp]: Oh yes. That's more efficient. Fixed.

File: src/crypto/hkdf_sha256_32.h [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+/** A rfc5869 HKDF implementation with HMAC_SHA256 and fixed key output length of 32 bytes (L=32) */
+class CHKDF_HMAC_SHA256_L32
+{
+private:
+    unsigned char m_prk[32];
+    static const size_t OUTPUT_SIZE = 32;
+
+public:
+    CHKDF_HMAC_SHA256_L32(const unsigned char* ikm, size_t ikmlen, const std::string& salt);
+    void Expand32(const std::string& info, unsigned char hash[OUTPUT_SIZE]);
```
[the remark this answers is not in the document]
Comment (jonasschnelli) [reply src/crypto/hkdf_sha256_32.h]: Thanks. Fixed.

**Comment by jonasschnelli** [comment]
Rebased

File: src/test/key_tests.cpp [hunk 3]
```diff
@@ -188,4 +188,14 @@ BOOST_AUTO_TEST_CASE(key_signature_tests)
     BOOST_CHECK(found_small);
 }
 
+BOOST_AUTO_TEST_CASE(key_key_negation)
+{
+    CKey key = DecodeSecret(strSecret1C);
```
Comment (sipa) [inline src/test/key_tests.cpp]: The test here could be a bit more extensive (for example, signing/verifying a fixed message with the corresponding private key, and checking that the private key roundtrips after 2 negations).

File: src/test/crypto_tests.cpp [hunk 4]
```diff
@@ -524,6 +525,16 @@ BOOST_AUTO_TEST_CASE(chacha20_testvector)
                  "fab78c9");
 }
 
+BOOST_AUTO_TEST_CASE(hkdf_hmac_sha256_l32_tests)
+{
+    // rfc5869 has no test vectors for HMAC_SHA256 length 32, the one below was created with python HKDF and compared against openssl
```
Comment (sipa) [inline src/test/crypto_tests.cpp]: I think shorter-length HMAC_HKDF can be obtained by truncating the output of a longer-length one. So you could take all of the RFC's test vectors and truncate them to 32 bytes?

Comment (jonasschnelli) [reply src/test/crypto_tests.cpp]: Good idea. Followed @sipa's advice and changed to the 3 rfc5869 SHA256 test vectors (truncated OKM to 32 bytes).

File: src/test/key_tests.cpp (hunk 3 above)
Comment (jonasschnelli) [reply src/test/key_tests.cpp]: Good point. Extended the test to cover the double negation signature comparison.

**Comment by sipa** [comment]
utACK acf3debc6a2f8b80a5de1e71bc658086c6de77b4

**Comment by jonasschnelli** [comment]
rebased

File: src/crypto/hkdf_sha256_32.cpp [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <crypto/hkdf_sha256_32.h>
+
+#include <assert.h>
+#include <string.h>
+
+CHKDF_HMAC_SHA256_L32::CHKDF_HMAC_SHA256_L32(const unsigned char* ikm, size_t ikmlen, const std::string& salt)
+{
+    CHMAC_SHA256(reinterpret_cast<const unsigned char*>(salt.c_str()), salt.size()).Write(ikm, ikmlen).Finalize(m_prk);
```
Comment (sipa) [inline src/crypto/hkdf_sha256_32.cpp]: Style nit: we don't usually use the C++ cast operators for primitive types (just `(const unsigned char*)salt.c_str()` works).

File: src/crypto/hkdf_sha256_32.cpp [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+}
+
+void CHKDF_HMAC_SHA256_L32::Expand32(const std::string& info, unsigned char hash[OUTPUT_SIZE])
+{
+    // expand a 32byte key (single round)
+    assert(info.size() <= 128);
+    static const unsigned char one[1] = {1};
+    CHMAC_SHA256(m_prk, 32).Write(reinterpret_cast<const unsigned char*>(info.data()), info.size()).Write(one, 1).Finalize(hash);
```
Comment (sipa) [inline src/crypto/hkdf_sha256_32.cpp]: Same here.

**Review by sipa — COMMENTED** [review]
utACK 65948ee3d41c33cd442a6572ab9e9d071499022d, just some nits.

**Comment by jonasschnelli** [comment]
Thanks for the review. Fixed the C++ cast nit.

**Comment by promag** [comment]
Restarted failed job.

File: src/crypto/hkdf_sha256_32.cpp [hunk 7]
```diff
@@ -0,0 +1,21 @@
+// Copyright (c) 2018 The Bitcoin Core developers
```
Comment (promag) [inline src/crypto/hkdf_sha256_32.cpp]: nit, 2019?

File: src/test/crypto_tests.cpp [hunk 8]
```diff
@@ -212,6 +213,22 @@ static void TestPoly1305(const std::string &hexmessage, const std::string &hexke
     BOOST_CHECK(tag == tagres);
 }
 
+static void TestHKDF_SHA256_32(const std::string &ikm_hex, const std::string &salt_hex, const std::string &info_hex, const std::string &okm_check_hex) {
+    std::vector<unsigned char> initial_key_material = ParseHex(ikm_hex);
+    std::vector<unsigned char> salt = ParseHex(salt_hex);
+    std::vector<unsigned char> info = ParseHex(info_hex);
+
+
```
Comment (promag) [inline src/test/crypto_tests.cpp]: nit, remove 2nd empty line.

File: src/test/crypto_tests.cpp [hunk 9]
```diff
@@ -212,6 +213,22 @@ static void TestPoly1305(const std::string &hexmessage, const std::string &hexke
     BOOST_CHECK(tag == tagres);
 }
 
+static void TestHKDF_SHA256_32(const std::string &ikm_hex, const std::string &salt_hex, const std::string &info_hex, const std::string &okm_check_hex) {
```
Comment (promag) [inline src/test/crypto_tests.cpp]: nit, space after `&`, not before.

**Review by promag — COMMENTED** [review]
some nits 👼

**Comment by laanwj** [comment]
utACK 8794a4b3ae4d34a4cd21a7dee9f694eef7726a4f
Don't want to hold this up on a few last-minute style nits and empty lines.

**Pull request merged**